internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,142 @@ import (
+	// InvocationContext is the context to be passed down the actor method.
+	InvocationContext() InvocationContext
+}
+
+// InvocationContext is passed to the actors on each method call..
+type InvocationContext interface {
+	// Review: for @go-filecoin, composition or inheritance?
+	Runtime() Runtime2
+	Caller() address.Address
+	StateHandler() ActorStateHandler
+
+	// Review: for @go-filecoin+@spec, renamed `ValueSupplied` to `ValueReceived`. The "supplied" is not clear by whom.
+	ValueReceived() types.AttoFIL
+	Balance() types.AttoFIL
+
+	LocalStorage() IpldStorage2
+}
 
-	CreateNewActor(addr address.Address, code cid.Cid, initalizationParams interface{}) error
+// initInvocationContext is a more powerfull context for the "init" actor.
+type initInvocationContext interface {
+	InvocationContext
+	// Review: for @spec, added an error that is not yet on the spec but should be
+	// Note: @jzimmerman: "Agreed, this should support an error return."
+	CreateActor(cid cid.Cid, addr address.Address, constructorParams MethodParams) error
+}
 
+// minerInvocationContext has some special sauce for the miner.
+// Review: for @go-filecoin, do we want custom interfaces for the miners instead of grouping a bunch fo methods on the InvocationContext?
+type minerInvocationContext interface {
 	Verifier() verification.Verifier
 }
 
+// ActorStateHandler handles the actor state, allowing actors to lock on the state.
+type ActorStateHandler interface {
+	// Take returns the `Cid` for the state and locks the state.
+	//
+	// Any futre calls to `Take` by other instances will block until `Release` or `UpdateRelease` is called.
+	// Review: for @spec, the spec impl of `Take` is not doing any blocking, just panic'ing on the second call on the SAME object instance (this objects are not kept around, so it is always a new instance..).
+	Take() cid.Cid
+	// UpdateRelease updates the actor state and releases the lock on it.
+	//
+	// No future calls to `Take` are allowed on this object.
+	// Review: for @spec, this method can currently be called without having called `Take` first.
+	UpdateRelease(newStateCID ActorSubstateCID)
+	// Release asserts that the state has not changed and releases the lock on it.
+	//
+	// No future calls to `Take` are allowed on this object.
+	// Review: for @spec, this method can currently be called without having called `Take` first.
+	// Review: for @spec, why is this method required?
+	Release(checkStateCID ActorSubstateCID)
+}
+
+// IpldStorage2 defines the storage module exposed to actors.
+type IpldStorage2 interface {
+	Put(interface{}) (cid.Cid, error)
+	Get(cid.Cid, interface{}) error
+}
+
+// CallerPattern checks if the caller matches the pattern.
+type CallerPattern interface {
+	IsMatch() bool
+}
+
+// Epoch is the current chain epoch.
+// Dragons: this is just to match names on the VM, Height got renamed to Epoch on the spec
+// TODO: rename types.BlockHeight to block.Epoch (issue: ???)
+type Epoch types.BlockHeight
```

Comment:
I *think* that we should be replacing go-filecoin's `types.BlockHeight` (which wraps bigint) with a newtype for `types.Uint64`, and then yes replacing with `Epoch` everywhere.

Related to https://github.com/filecoin-project/specs/issues/615, we need our own issue for that busywork.

Reply:
SGTM